Support pyrax.auth_with_token in rackspace modules

kouk opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
When automating ansible or running in a less than fully trusted
environment it's desirable to use the [scoped and short-lived
auth token](https://developer.rackspace.com/docs/cloud-servers/v2/getting-started/authenticate/)
that rackspace supports. For this we should optionally be able to use `pyrax.auth_with_token` instead of the credentials file.

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request


##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
rax

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.5.0 (rax_auth_token 52daaa08db) last updated 2017/09/11 12:16:40 (GMT +300)
  config file = None
  configured module search path = [u'/Users/kouk/code/ansible/library']
  ansible python module location = /Users/kouk/code/ansible/lib/ansible
  executable location = /Users/kouk/code/ansible/bin/ansible
  python version = 2.7.13 (default, Jul 18 2017, 09:17:00) [GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.42)]
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

This pr adds support for retrieving the auth token either from the
rax modules arguments or from the environment. It also uses the tenant
name since it is required when authenticating via auth token.

kouk committed:
Support pyrax.auth_with_token in rackspace modules

When automating ansible or running in a less than fully trusted
environment it's desirable to use the [scoped and short-lived
auth token](https://developer.rackspace.com/docs/cloud-servers/v2/getting-started/authenticate/)
that rackspace supports. For this we can use `pyrax.auth_with_token`.

This pr adds support for retrieving the auth token either from the
rax modules arguments or from the environment. It also uses the tenant
name since it is required when authenticating via auth token.

kouk commented:
bot_status

kouk commented:
any news on this?

kouk commented:
bump?

mattclay commented on lib/ansible/module_utils/rax.py:
```diff
@@ -244,6 +244,7 @@ def rax_argument_spec():
     return dict(
         api_key=dict(type='str', aliases=['password'], no_log=True),
         auth_endpoint=dict(type='str'),
+        auth_token=dict(type='str'),
```
This needs to be documented in the rackspace docs fragment.

kouk replied:
Thanks, I missed that. I've pushed the necessary changes (waiting for CI).

kouk replied:
CI passed.

mattclay replied:
Since this is a credential `no_log=True` should be set here.

kouk commented:
bot_status

kouk commented:
@mattclay could you take another look at this? Thank you in advance.

kouk commented:
@mattclay all done from my side AFAICS. Am I missing something?

mattclay commented:
@kouk The changes look OK to me, but I'm not a user of the Rackspace modules. I recommend finding another contributor/user of the Rackspace modules to review and test.

kouk commented:
@dagwieers @abadger @immanetize @sivel could one of you perhaps help review this?

Erethon commented:
We've been using this in production for quite a while now and it's been working fine. cc @mattclay

kouk commented:
bot_status

kouk commented:
@mattclay could you suggest a way forward for this? Should I reach out on a mailing list maybe to find someone else to review besides @Erethon and you?

mattclay commented on lib/ansible/module_utils/rax.py:
```diff
@@ -305,6 +307,8 @@ def setup_rax_module(module, rax_module, region_required=True):
         module.fail_json(msg='Unable to load %s' % e.message)
 
+    token = token or os.environ.get('OS_TOKEN')
+    tenant_name = tenant_name or os.environ.get('OS_PROJECT_NAME')
```
These environment variable names should probably be `RAX_TOKEN` and `RAX_TENANT_NAME` to match the existing `RAX_` prefixed environment variables used by this module_util.

Also note the change from `PROJECT` to `TENANT` to match the terminology used in the API docs.

mattclay commented on lib/ansible/module_utils/rax.py:
```diff
@@ -276,6 +277,7 @@ def setup_rax_module(module, rax_module, region_required=True):
     tenant_name = module.params.get('tenant_name')
     username = module.params.get('username')
     verify_ssl = module.params.get('verify_ssl')
+    token = module.params.get('auth_token')
```
Use `auth_token` to match the parameter name.

mattclay commented on lib/ansible/utils/module_docs_fragments/rackspace.py:
```diff
@@ -74,6 +74,11 @@ class ModuleDocFragment(object):
+    description:
+      - A pre-generated authentication token (requires I(tenant_name) or C(OS_PROJECT_NAME))
+        see U(https://developer.rackspace.com/docs/cloud-identity/v2/api-reference/token-operations/).
+    version_added: 2.6.0
```
Update this to `2.8` since that will be the upcoming release this could be included in.

mattclay reviewed (dismissed):
I've reviewed the PR and requested a few minor changes. I also restarted CI to make sure the code passes our current tests (we're always adding new tests).

Once the changes are made, and you get an updated review and feedback from a community member that uses the RAX modules, this should be ready to merge.

kouk commented:
@mattclay @Erethon whenever you have a chance, please check the latest changes as I believe I've addressed the issues. I can squash the commits into one if you're satisfied with them.

kouk commented:
@Erethon  can you take a look?

Erethon commented:
LGTM, been using it for a while with the new variable names without any issues.

kouk commented:
ping @mattclay

kouk commented:
@mattclay @Erethon i rebased onto devel but it seems an apparently unrelated test in shippable finished with an error. Can we retry the build?
